Fix #28144: Include score metadata (Author, Subject, Composer, Arranger, Translator, Lyricist, and Copyright) as PDF metadata on export

YueZheng-Sea-angle opened the pull request:
Resolves: #28144   <!-- Replace `NNNNN` with a GitHub issue number, or a direct link if the issue is not on GitHub -->

Implemented embedding score metadata into PDF files as discussed in issue #28144. Added mappings for seven metadata fields: **Author, Subject, Composer, Arranger, Translator, Lyricist, and Copyright**. The implementation has been locally compiled, run, and tested.
<!-- Add a short description of and motivation for the changes here -->

<!-- Replace `[ ]` with `[x]` to fill the checkboxes below -->

- [x] I signed the [CLA](https://musescore.org/en/cla)
- [x] The title of the PR describes the problem it addresses
- [x] Each commit's message describes its purpose and effects, and references the issue it resolves
- [x] If changes are extensive, there is a sequence of easily reviewable commits
- [x] The code in the PR follows [the coding rules](https://github.com/musescore/MuseScore/wiki/CodeGuidelines)
- [x] There are no unnecessary changes
- [x] The code compiles and runs on my machine, preferably after each commit individually
- [ ] I created a unit test or vtest to verify the changes I made (if applicable)

Jojo-Schmitz commented on src/project/qml/MuseScore/Project/internal/Export/PdfSettingsPage.qml:
```diff
@@ -26,48 +26,63 @@ import Muse.UiComponents 1.0
+
+    CheckBox {
+width:  parent.width
+text:   qsTrc("project/export", "Embed metadata")
```
and wrong indent here

Jojo-Schmitz commented on src/project/qml/MuseScore/Project/internal/Export/PdfSettingsPage.qml:
```diff
@@ -26,48 +26,63 @@ import Muse.UiComponents 1.0
+        navigation.panel:root.navigationPanel
+        navigation.row:root.navigationOrder + 3
+
+checked: root.model.pdfEmbeddedMetadata
```
and here

cbjeukendrup commented on src/importexport/imagesexport/internal/pdfwriter.cpp:
```diff
@@ -147,11 +149,107 @@ Ret PdfWriter::writeList(const INotationPtrList& notations, io::IODevice& destin
+// Extract project metadata from current project or notation fallback
+ProjectMeta PdfWriter::getProjectMetadata(INotationPtr notation) const
+{
+    auto project = globalContext()->currentProject();
```
It would be great if we could avoid the dependency on `globalContext` here. The notation being exported should not necessarily correspond to the currently opened project. 

I see two ways:
* Move the `metaInfo` method to `INotation` (and perhaps let `NotationProject` still have a `metaInfo` that just calls `notation->metaInfo`). 
* Give `INotation` a pointer to its `NotationProject`, and retrieve the project from the `notation`. Perhaps that's a good idea anyway, and shouldn't be difficult. Note that it should be a raw pointer or a weak pointer, but not a shared pointer, because otherwise the MasterNotation and NotationProject will keep each other alive (a "retain cycle", which is a kind of memory leak).

cbjeukendrup commented on src/project/view/exportdialogmodel.h:
```diff
@@ -69,6 +69,8 @@ class ExportDialogModel : public QAbstractListModel, public muse::async::Asyncab
     Q_PROPERTY(
         bool pdfTransparentBackground READ pdfTransparentBackground WRITE setPdfTransparentBackground NOTIFY pdfTransparentBackgroundChanged)
+    Q_PROPERTY(
+        bool pdfEmbeddedMetadata READ pdfEmbeddedMetadata WRITE setPdfEmbeddedMetadata NOTIFY pdfEmbeddedMetadataChanged)
```
Nitpick: `pdfEmbedMetadata` seems more natural to me than `pdfEmbeddedMetadata`

Jojo-Schmitz commented on src/importexport/imagesexport/internal/pdfwriter.cpp:
```diff
@@ -147,11 +149,107 @@ Ret PdfWriter::writeList(const INotationPtrList& notations, io::IODevice& destin
+    QString currentDateTime = QDateTime::currentDateTime().toString(Qt::ISODate);
+
+    // Populate XMP template with project metadata
+    QString xmpData = xmpTemplate
+                      .arg(escapeXml(meta.title))
+                      .arg(escapeXml(creator))
+                      .arg(escapeXml(meta.subtitle))
+                      .arg(escapeXml(meta.copyright))
+                      .arg(currentDateTime)
+                      .arg(escapeXml(author))
+                      .arg(escapeXml(meta.composer))
+                      .arg(escapeXml(meta.arranger))
+                      .arg(escapeXml(meta.translator))
+                      .arg(escapeXml(meta.lyricist));
```
Better indeed, if, like in thos case, the args are all of the same type

Jojo-Schmitz commented on src/importexport/imagesexport/internal/pdfwriter.cpp:
```diff
@@ -147,11 +149,107 @@ Ret PdfWriter::writeList(const INotationPtrList& notations, io::IODevice& destin
+        xmlns:pdfx="http://ns.adobe.com/pdfx/1.3/">
+      <dc:title>%1</dc:title>
+      <dc:creator>%2</dc:creator>
+      <pdf:Author>%6</pdf:Author>
```
Maybe order them the same way as below in `xmpTemplate`?

YueZheng-Sea-angle commented:
I have addressed all the review points mentioned above.

Additionally, I resolved the merge conflicts caused by PR #30213. Since that PR already introduced a new checkbox, adding another one here would cause the `Open destination folder on export` option to be obscured by the `Export...` and `Cancel` buttons. As an improvement, I have slightly increased the height of the PDF export dialog to prevent this UI issue.

I also noted the recent discussions in issue #28144 regarding the addition of a checkbox. However, based on the previous guidance from the maintainers, I have not implemented any changes related to that discussion for now.

Please let me know if any modifications are needed or what the next steps should be.
<img width="1919" height="1019" alt="change" src="https://github.com/user-attachments/assets/0f5e8e07-64d6-491c-9994-de7f73723726" />

cbjeukendrup commented:
@YueZheng-Sea-angle may I ask you to take care of the above? (i.e. double-check which values are put in which metadata fields, and optimise that to work well with PDF viewers)

YueZheng-Sea-angle commented:
Actually, I have spent months testing with different metadata display methods, yet the results remain inconsistent—even though certain websites have managed it perfectly from the very beginning. My tests show that standard fields work in Acrobat while complex ones appear in advanced XMP, which I consider to be OK. It is impossible to guess how to satisfy the specific standards of different software without clear requirements.
Additionaly, the Title and Composer logic follows forscore's official documentation. I'm uncertain how to improve it. Howewer, I'm sorry that I'm not in a position to purchase it.
With this framework, anyone who identifies the exact metadata definitions in the future can resolve compatibility quickly.
Hope the code helps future contributors. Thanks for helping and reviewing!
